Source/ExecutionEngine/ConcurrentToSequentialWriteManager.cs

```diff
@@ -50,9 +53,10 @@ class SubWriter : WriterWrapper {
     }
 
     public string SetTargetAndGetBuffer(TextWriter newTarget) {
-      var result = buffering ? ((StringWriter)target).ToString() : "";
-      buffering = false;
+      var oldTarget = target;
       target = newTarget;
+      var result = buffering ? ((StringWriter)oldTarget).ToString() : "";
+      buffering = false;
       return result;
```

Comment:
Nice catch. This is a subtle issue!

Reply:
Great catch. I'm wondering if it is still possible to have another race condition however: Two threads stepping in this code could both assign `oldTarget` together, and then assign `target` together, and one StringWriter would not be rendered.
I think the whole function should be locked someway.

Reply:
The calls to this method are inside a lock

Reply:
I think the "right" way to solve these sort of issues is `Interlocked.Exchange`, but I don't actually understand the root cause, here.

If this method is called within an exclusive lock, how can this change fix an issue?  And if not, why not just lock the whole method body on `myLock`?

Reply:
The issue was that we had:
```
var result = buffering ? ((StringWriter)target).ToString() : "";
...
target = newTarget;
```

So something could be written to `target` after the `var result` line but before the `target = newTarget` line.

Reply:
Yes but do you have a guarantee that writes to these two variables result and target are done in order?  Otherwise the fix is not guaranteed to help, right?